File: src/caffeine_query_language/parser_ffi.erl
%% Fast byte-level helpers paralleling parser_ffi.mjs. Erlang strings are UTF-8
%% binaries; for ASCII tokens (operators, parens, keyword bytes) byte position
%% equals UTF-8 codepoint position, so binary:at / binary:part are correct.
-module(parser_ffi).
-export([code_unit_at/2, substring_equals_at/3, code_unit_length/1, slice_codeunits/3]).

code_unit_at(S, I) when is_binary(S), is_integer(I), I >= 0 ->
    case byte_size(S) > I of
        true -> binary:at(S, I);
        false -> -1
    end;
code_unit_at(_, _) -> -1.

substring_equals_at(Haystack, Pos, Needle) when is_binary(Haystack), is_binary(Needle), is_integer(Pos), Pos >= 0 ->
    NeedleSize = byte_size(Needle),
    case byte_size(Haystack) >= Pos + NeedleSize of
        true -> binary:part(Haystack, Pos, NeedleSize) =:= Needle;
        false -> false
    end;
substring_equals_at(_, _, _) -> false.

code_unit_length(S) when is_binary(S) -> byte_size(S);
code_unit_length(_) -> 0.

slice_codeunits(S, Start, Len) when is_binary(S), is_integer(Start), is_integer(Len), Start >= 0, Len > 0 ->
    Total = byte_size(S),
    case Start >= Total of
        true -> <<>>;
        false ->
            ClampedLen = min(Len, Total - Start),
            binary:part(S, Start, ClampedLen)
    end;
slice_codeunits(_, _, _) -> <<>>.

File: src/caffeine_query_language/parser_ffi.mjs
// Fast codeunit-level access used by parser.gleam to avoid the Intl.Segmenter
// machinery in gleam_stdlib's string.slice / pop_grapheme. Those grapheme-safe
// helpers construct a new Segmenter iterator per call and walk it from index 0,
// turning the parser's per-position scans into O(N^2). For pure-ASCII tokens
// (operators, parens, keywords like "per") UTF-16 codeunit access is correct
// and orders of magnitude faster.

// Returns the UTF-16 code unit at index i, or -1 if i is out of bounds.
export function code_unit_at(s, i) {
  if (i < 0 || i >= s.length) return -1;
  return s.charCodeAt(i);
}

// Byte-substring equality: true when s.substring(pos, pos+needle.length) === needle.
// Uses substring (no Segmenter) and short-circuits on length check.
export function substring_equals_at(haystack, pos, needle) {
  const end = pos + needle.length;
  if (pos < 0 || end > haystack.length) return false;
  return haystack.substring(pos, end) === needle;
}

// Codeunit count — matches the indexing space used by code_unit_at /
// substring_equals_at / slice_codeunits. Crucially this is NOT grapheme
// count (which gleam_stdlib's string.length returns).
export function code_unit_length(s) {
  return s.length;
}

// Slice by codeunit range. Returns "" if start is out of bounds; clamps len.
export function slice_codeunits(s, start, len) {
  if (start < 0 || start >= s.length || len <= 0) return "";
  return s.substring(start, start + len);
}
